source/slang/slang-preprocessor.cpp

```diff
@@ -1,198 +1,648 @@
+        //
+        auto top = m_top;
+        for(;;)
+        {
+            SLANG_ASSERT(top);
+            auto token = top->peekToken();
+            if( token.type != TokenType::EndOfFile )
+                return token;
+
+            auto parent = top->getParent();
+            if(parent)
+            {
+                top = parent;
+                continue;
+            }
+
+            return token;
+        }
+    }
+
+        /// Return type of the token that `peekToken()` will return
+    TokenType peekTokenType()
+    {
+        return peekToken().type;
+    }
+
+        /// Return location of the token that `peekToken()` will return
+    SourceLoc peekLoc()
+    {
+        return peekToken().loc;
+    }
+
+        /// Skip over all whitespace tokens in the input stream(s) to arrive at the next non-whitespace token
+    void skipAllWhitespace()
+    {
+        for( ;;)
+        {
+            switch(peekTokenType())
+            {
+            default:
+                return;
+
+                // Note: We expect `NewLine` to be the only case of whitespace we
+                // encounter right now, because all the other cases will have been
+                // filtered out by the `LexerInputStream`.
+                //
+            case TokenType::NewLine:
+            case TokenType::WhiteSpace:
+            case TokenType::BlockComment:
+            case TokenType::LineComment:
+                readToken();
+                break;
+            }
+        }
+    }
+
+        /// Get the top stream of the input stack
+    InputStream* getTopStream()
+    {
+        return m_top;
+    }
+
+        /// Get the input stream that the next token would come from
+    InputStream* getNextStream()
+    {
+        SLANG_ASSERT(m_top);
+        auto top = m_top;
+        for(;;)
+        {
+            auto token = top->peekToken();
+            if( token.type != TokenType::EndOfFile )
+                return top;
+
+            auto parent = top->getParent();
+            if(parent)
+            {
+                top = parent;
+                continue;
+            }
+
+            return top;
+        }
+    }
+
+private:
+            /// The top of the stack of input streams
+    InputStream* m_top = nullptr;
 };
 
-// An enumeration for the different types of macros
-enum class PreprocessorMacroFlavor
+// Another (relatively) simple case of an input stream is one that reads
+// tokens directly from the lexer.
+//
+// It might seem like we could simplify things even further by always lexing
+// a file into tokens first, and then using the earlier input-stream cases
+// for pre-tokenized input. The main reason we don't use that strategy is
+// that when dealing with preprocessor conditionals we will often want to
+// suppress diagnostic messages coming from the lexer when inside of disabled
+// conditional branches.
+//
+// TODO: We might be able to simplify the logic here by having the lexer buffer
```

Comment:
In my mind representing the diagnostics as tokens is a kind of "buffering up" - and the most likely kind we'd implement. The main problem is that if you follow that line of reasoning you soon find that the preprocessor is the only logical place to own the code that emits the diagnostics: the preprocessor knows what code is being skipped (and doesn't want diagnostics) and also controls the policy for when tokens might get duplicated or elided as part of macro expansion (so that trying to emit diagnostics *after* preprocessing would probably be a Bad Idea).

*If* we had a more flexible system where we could customize "diagnostic sink" types, then it might make sense for the preprocessor to initialize the lexer with a custom diagnostic sink type that it can use to turn emission of diagnostics on/off. That would allow the code in the lexer to always emit the diagnostic to its stream, but then the policy on the stream itself would determine what actually makes it out to the user.